File: astronomical_sundial.rb
# astronomical_sundial.rb
#!/usr/bin/env ruby
require 'optparse'
require 'date'

DEG2RAD = Math::PI / 180.0
RAD2DEG = 180.0 / Math::PI

def julian_day(dt)
  year = dt.year
  month = dt.month
  day = dt.day + dt.hour/24.0 + dt.min/1440.0 + dt.sec/86400.0
  if month <= 2
    year -= 1
    month += 12
  end
  a = (year / 100).to_i
  b = 2 - a + (a / 4).to_i
  (365.25 * (year + 4716)).to_i + (30.6001 * (month + 1)).to_i + day + b - 1524.5
end

def solar_declination(day_of_year)
  23.44 * DEG2RAD * Math.sin((284 + day_of_year) * 360 * DEG2RAD / 365)
end

def equation_of_time(day_of_year)
  b = (360.0 / 365) * (day_of_year - 81)
  b_rad = b * DEG2RAD
  9.87 * Math.sin(2 * b_rad) - 7.53 * Math.cos(b_rad) - 1.5 * Math.sin(b_rad)
end

def sun_position(lat_deg, lon_deg, dt)
  lat_rad = lat_deg * DEG2RAD
  day_of_year = dt.yday
  dec_rad = solar_declination(day_of_year)
  eot = equation_of_time(day_of_year)
  hour_utc = dt.hour + dt.min/60.0 + dt.sec/3600.0
  solar_time = hour_utc + (4 * lon_deg) / 60.0 + eot / 60.0
  ha_rad = (solar_time - 12) * 15 * DEG2RAD

  alt_rad = Math.asin(Math.sin(lat_rad) * Math.sin(dec_rad) + Math.cos(lat_rad) * Math.cos(dec_rad) * Math.cos(ha_rad))
  alt_deg = alt_rad * RAD2DEG

  azi_rad = Math.atan2(-Math.sin(ha_rad) * Math.cos(dec_rad),
                       Math.sin(dec_rad) * Math.cos(lat_rad) -
                       Math.cos(dec_rad) * Math.sin(lat_rad) * Math.cos(ha_rad))
  azi_deg = (azi_rad * RAD2DEG) % 360.0

  cos_ha_sunrise = -Math.tan(lat_rad) * Math.tan(dec_rad)
  ha_sunrise = if cos_ha_sunrise < -1
                 Math::PI
               elsif cos_ha_sunrise > 1
                 0
               else
                 Math.acos(cos_ha_sunrise)
               end
  day_length = ha_sunrise * 2 / (Math::PI / 12)
  noon = 12.0 - lon_deg/15.0 - eot/60.0
  sunrise = noon - day_length/2
  sunset = noon + day_length/2

  {
    altitude: alt_deg,
    azimuth: azi_deg,
    solar_time: solar_time,
    eot: eot,
    declination: dec_rad * RAD2DEG,
    sunrise: sunrise,
    sunset: sunset,
    day_length: day_length
  }
end

def draw_sundial(azimuth_deg)
  dir_names = ['N', 'NE', 'E', 'SE', 'S', 'SW', 'W', 'NW']
  idx = (azimuth_deg / 45).round % 8
  shadow_dir = dir_names[idx]
  lines = [
    "      N",
    "      |",
    "  W---+---E",
    "      |",
    "      S",
    "\nShadow direction: #{shadow_dir} (#{azimuth_deg.round(1)}°)"
  ]
  lines.join("\n")
end

def format_time(hours)
  h = hours.to_i % 24
  m = ((hours - h) * 60).to_i
  sprintf("%02d:%02d", h, m)
end

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: astronomical_sundial.rb [options]"
  opts.on("--lat LAT", Float, "Latitude (positive North)") { |v| options[:lat] = v }
  opts.on("--lon LON", Float, "Longitude (positive East)") { |v| options[:lon] = v }
  opts.on("--tz TZ", Float, "Timezone offset (hours from UTC)") { |v| options[:tz] = v }
  opts.on("--date DATE", "YYYY-MM-DD") { |v| options[:date] = v }
  opts.on("--time TIME", "HH:MM") { |v| options[:time] = v }
  opts.on("--dial-only", "Show only the dial") { options[:dial_only] = true }
end.parse!

now = DateTime.now
dt = now
if options[:date]
  dt = DateTime.parse(options[:date])
end
if options[:time]
  h, m = options[:time].split(':').map(&:to_i)
  dt = DateTime.new(dt.year, dt.month, dt.day, h, m, 0, dt.zone)
else
  dt = DateTime.new(dt.year, dt.month, dt.day, now.hour, now.min, 0, dt.zone)
end
# Convert to UTC
dt = dt.new_offset(0)

lat = options[:lat] || 0.0
lon = options[:lon] || 0.0
tz = options[:tz] || 0.0

if options[:dial_only]
  data = sun_position(lat, lon, dt)
  puts draw_sundial(data[:azimuth])
  exit
end

data = sun_position(lat, lon, dt)
alt = data[:altitude]
azi = data[:azimuth]
solar_time = data[:solar_time]
eot = data[:eot]
dec = data[:declination]
sunrise = data[:sunrise] + tz
sunset = data[:sunset] + tz
day_len = data[:day_length]
day_len_h = day_len.to_i
day_len_m = ((day_len - day_len_h) * 60).to_i

lat_str = "#{lat.abs.round(2)}°#{lat >= 0 ? 'N' : 'S'}"
lon_str = "#{lon.abs.round(2)}°#{lon >= 0 ? 'E' : 'W'}"
tz_sign = tz >= 0 ? '+' : '-'
local_time = dt + tz/24.0

puts "\n☀️ Astronomical Sundial"
puts "Location: #{lat_str}, #{lon_str}"
puts "Date: #{dt.strftime('%Y-%m-%d %H:%M')} (UTC#{tz_sign}#{tz.abs.round(1)})"
puts "Local Time: #{local_time.strftime('%H:%M')}"
puts "\nSolar Declination: #{dec.round(1)}°"
puts "Equation of Time: #{eot.round(1)} min"
solar_h = solar_time.to_i
solar_m = ((solar_time - solar_h) * 60).to_i
puts "Solar Time: #{sprintf('%02d:%02d', solar_h, solar_m)}"
puts "\nSolar Altitude: #{alt.round(1)}°"
puts "Solar Azimuth: #{azi.round(1)}°"
puts "\nSunrise: #{format_time(sunrise)} | Sunset: #{format_time(sunset)}"
puts "Day length: #{day_len_h}h #{day_len_m}m"
puts "\n" + draw_sundial(azi)

File: astronomical_sundial.go
// astronomical_sundial.go
package main

import (
	"flag"
	"fmt"
	"math"
	"time"
)

const (
	DEG2RAD = math.Pi / 180.0
	RAD2DEG = 180.0 / math.Pi
)

type SolarData struct {
	Altitude    float64
	Azimuth     float64
	SolarTime   float64
	EOT         float64
	Declination float64
	HA          float64
	Sunrise     float64
	Sunset      float64
	DayLength   float64
}

func julianDay(t time.Time) float64 {
	year := t.Year()
	month := int(t.Month())
	day := float64(t.Day()) + float64(t.Hour())/24.0 + float64(t.Minute())/1440.0 + float64(t.Second())/86400.0
	if month <= 2 {
		year--
		month += 12
	}
	A := year / 100
	B := 2 - A + A/4
	return float64(int(365.25*float64(year+4716))) + float64(int(30.6001*float64(month+1))) + day + float64(B) - 1524.5
}

func solarDeclination(dayOfYear int) float64 {
	return 23.44 * DEG2RAD * math.Sin((284+dayOfYear)*360*DEG2RAD/365)
}

func equationOfTime(dayOfYear int) float64 {
	B := (360.0 / 365) * float64(dayOfYear-81)
	B_rad := B * DEG2RAD
	return 9.87*math.Sin(2*B_rad) - 7.53*math.Cos(B_rad) - 1.5*math.Sin(B_rad)
}

func sunPosition(latDeg, lonDeg float64, t time.Time) SolarData {
	latRad := latDeg * DEG2RAD
	dayOfYear := t.YearDay()
	decRad := solarDeclination(dayOfYear)
	eot := equationOfTime(dayOfYear)
	hourUTC := float64(t.Hour()) + float64(t.Minute())/60.0 + float64(t.Second())/3600.0
	solarTime := hourUTC + (4*lonDeg)/60.0 + eot/60.0
	haRad := (solarTime - 12) * 15 * DEG2RAD

	altRad := math.Asin(math.Sin(latRad)*math.Sin(decRad) + math.Cos(latRad)*math.Cos(decRad)*math.Cos(haRad))
	altDeg := altRad * RAD2DEG

	aziRad := math.Atan2(-math.Sin(haRad)*math.Cos(decRad),
		math.Sin(decRad)*math.Cos(latRad)-math.Cos(decRad)*math.Sin(latRad)*math.Cos(haRad))
	aziDeg := math.Mod(aziRad*RAD2DEG+360, 360)

	// Sunrise/sunset
	cosHASunrise := -math.Tan(latRad) * math.Tan(decRad)
	var haSunrise float64
	if cosHASunrise < -1 {
		haSunrise = math.Pi
	} else if cosHASunrise > 1 {
		haSunrise = 0
	} else {
		haSunrise = math.Acos(cosHASunrise)
	}
	dayLength := haSunrise * 2 / (math.Pi / 12)
	noon := 12.0 - lonDeg/15.0 - eot/60.0
	sunrise := noon - dayLength/2
	sunset := noon + dayLength/2

	return SolarData{
		Altitude:    altDeg,
		Azimuth:     aziDeg,
		SolarTime:   solarTime,
		EOT:         eot,
		Declination: decRad * RAD2DEG,
		HA:          haRad * RAD2DEG,
		Sunrise:     sunrise,
		Sunset:      sunset,
		DayLength:   dayLength,
	}
}

func drawSundial(azimuthDeg float64) string {
	dirNames := []string{"N", "NE", "E", "SE", "S", "SW", "W", "NW"}
	idx := int(math.Round(azimuthDeg/45)) % 8
	shadowDir := dirNames[idx]
	lines := []string{
		"      N",
		"      |",
		"  W---+---E",
		"      |",
		"      S",
		fmt.Sprintf("\nShadow direction: %s (%.1f°)", shadowDir, azimuthDeg),
	}
	return strings.Join(lines, "\n")
}

func formatTime(hours float64) string {
	h := int(hours) % 24
	m := int((hours - float64(h)) * 60)
	return fmt.Sprintf("%02d:%02d", h, m)
}

func main() {
	var (
		lat      = flag.Float64("lat", 0.0, "Latitude (degrees North)")
		lon      = flag.Float64("lon", 0.0, "Longitude (degrees East)")
		tz       = flag.Float64("tz", 0.0, "Timezone offset (hours from UTC)")
		dateStr  = flag.String("date", "", "YYYY-MM-DD")
		timeStr  = flag.String("time", "", "HH:MM")
		dialOnly = flag.Bool("dial-only", false, "Show only the dial")
	)
	flag.Parse()

	now := time.Now().UTC()
	t := now
	if *dateStr != "" {
		d, _ := time.Parse("2006-01-02", *dateStr)
		t = time.Date(d.Year(), d.Month(), d.Day(), t.Hour(), t.Minute(), 0, 0, time.UTC)
	}
	if *timeStr != "" {
		parsed, _ := time.Parse("15:04", *timeStr)
		t = time.Date(t.Year(), t.Month(), t.Day(), parsed.Hour(), parsed.Minute(), 0, 0, time.UTC)
	}

	if *dialOnly {
		data := sunPosition(*lat, *lon, t)
		fmt.Print(drawSundial(data.Azimuth))
		return
	}

	data := sunPosition(*lat, *lon, t)
	alt := data.Altitude
	azi := data.Azimuth
	solarTime := data.SolarTime
	eot := data.EOT
	dec := data.Declination
	sunrise := data.Sunrise + *tz
	sunset := data.Sunset + *tz
	dayLen := data.DayLength
	dayLenH := int(dayLen)
	dayLenM := int((dayLen - float64(dayLenH)) * 60)

	latStr := fmt.Sprintf("%.2f°%c", math.Abs(*lat), 'N')
	if *lat < 0 {
		latStr = fmt.Sprintf("%.2f°%c", math.Abs(*lat), 'S')
	}
	lonStr := fmt.Sprintf("%.2f°%c", math.Abs(*lon), 'E')
	if *lon < 0 {
		lonStr = fmt.Sprintf("%.2f°%c", math.Abs(*lon), 'W')
	}
	tzSign := '+'
	if *tz < 0 {
		tzSign = '-'
	}
	localTime := t.Add(time.Duration(*tz * float64(time.Hour)))

	fmt.Printf("\n☀️ Astronomical Sundial\n")
	fmt.Printf("Location: %s, %s\n", latStr, lonStr)
	fmt.Printf("Date: %s (UTC%c%.1f)\n", t.Format("2006-01-02 15:04"), tzSign, math.Abs(*tz))
	fmt.Printf("Local Time: %s\n", localTime.Format("15:04"))
	fmt.Printf("\nSolar Declination: %+.1f°\n", dec)
	fmt.Printf("Equation of Time: %+.1f min\n", eot)
	solarH := int(solarTime)
	solarM := int((solarTime - float64(solarH)) * 60)
	fmt.Printf("Solar Time: %02d:%02d\n", solarH, solarM)
	fmt.Printf("\nSolar Altitude: %.1f°\n", alt)
	fmt.Printf("Solar Azimuth: %.1f°\n", azi)
	fmt.Printf("\nSunrise: %s | Sunset: %s\n", formatTime(sunrise), formatTime(sunset))
	fmt.Printf("Day length: %dh %dm\n", dayLenH, dayLenM)
	fmt.Println("\n" + drawSundial(azi))
}
